Shortcut Cast.toNumber if given a number

mzgoddard opened the pull request:
### Resolves

Close #1705 

### Proposed Changes

Check if `typeof value === 'number'` and do a separate isNaN check to confirm its not NaN.

### Reason for Changes

If toNumber is called on a number avoiding passing the number to Number() can provide a small performance improvement.

### Numbers

This change alone does not improve the blocks per second or fps in Scratch due to how the Scratch VM runs. Scratch iterates all threads before checking if it has used too much time instead of checking after each thread. I mention the later because that would be more likely to show a difference in blocks per second or fps (and we don't use that method because it would make projects nondeterministic).

With that in mind for this set of performance numbers I've recorded the percent of time spent in Cast.toNumber with browser inspector performance panels.

Together with other changes this will help speed up Scratch.

|  Project, Platform | before (% of time) | after (% of time) | Change |
|  ------ | ------ | ------ | ------ |
|  89811578, 2017 MBP 15" Chrome | 1.50% | 1.04% | 0.46% |
|  89811578, 2017 MBP 15" Firefox | 1.79% | 1.16% | 0.63% |
|  89811578, 2017 MBP 15" Safari | 2.00% | 2.00% | 0.00% |
|  89811578, 2017 SMS Chromebook | 1.58% | 0.90% | 0.68% |
|  89811578, 2018 Raspberry Pi 3B+ | 1.33% | 1.12% | 0.21% |

mzgoddard committed:
check if toNumber is given a number and shortcut if so

If toNumber is called on a number avoiding passing the number to Number
can provide a small performance improvement.

kchadha reviewed (approved):
LGTM!

joker314 commented on src/util/cast.js:
```diff
@@ -20,6 +20,17 @@ class Cast {
+        if (typeof value === 'number') {
+            // Scratch treats NaN as 0, when needed as a number.
+            // E.g., 0 + NaN -> 0.
+            if (isNaN(value)) {
```
Is `Number.isNaN(value)` more efficient for a similar reason? `isNaN` still implicitly coerces to a number.

kchadha replied:
@joker314, good catch!

mzgoddard replied:
@joker314 That's a great point.

I made the change (with back compat support in mind for Internet Explorer which does not provide Number.isNaN, in which case self.isNaN is a safe backup) and recorded some new numbers.

|  Project, Platform | before (% of time) | after (% of time) | Change |
|  ------ | ------ | ------ | ------ |
|  89811578, 2017 MBP 15" Chrome | 1.50% | 0.93% | 0.57% |
|  89811578, 2017 MBP 15" Firefox | 1.79% | 1.24% | 0.55% |
|  89811578, 2017 MBP 15" Safari | 2.00% | 1.40% | 0.60% |
|  89811578, 2017 SMS Chromebook | 1.58% | 0.90% | 0.68% |
|  89811578, 2018 Raspberry Pi 3B+ | 1.33% | 0.95% | 0.38% |

Chrome sees a minor change with it.
Looks like I could have recorded my Firefox numbers better.
And Safari sees an improvement now.

joker314 replied:
I figure I'd construct a table to show the change for just this commit. (I think I worked this out correctly?)

<table>
<tr><th>Project, Platform</th><th>isNaN (%)</th><th>Number.isNaN (%)</th><th>Change of Change (%)</tr>
<tr><td>89811578, 2017 MBP 15" Chrome</td><td>0.46</td><td>0.57</td><td><strong>+0.11</strong></td></tr>

<tr><td>89811578, 2017 MBP 15" Firefox</td><td>0.63</td><td>0.55</td><td><strong>-0.08</strong></td></tr>

<tr><td>89811578, 2017 MBP 15" Safari</td><td>0.00</td><td>0.60</td><td><strong>+0.60</strong></td></tr>

<tr><td>89811578, 2017 SMS Chromebook</td><td>0.68</td><td>0.68</td><td><strong>±0.00</strong></td></tr>

<tr><td>89811578, 2018 Raspberry Pi 3B+</td><td>0.21</td><td>0.38</td><td><strong>+0.17</strong></td></tr>
</table>

mzgoddard committed:
use Number.isNaN in Cast.toNumber

Number.isNaN does not coerce and may help performance since we either
do not need to coerce the value or already have with Number.

kchadha reviewed (changes requested):
@mzgoddard, looks like lint is failing because of the naming convention for `Number_isNaN`

mzgoddard commented:
Lint fixed.

kchadha commented on src/util/cast.js:
```diff
@@ -1,5 +1,13 @@
+/**
+ * Store and possibly polyfill Number.isNaN. Number.isNaN can save time over
+ * self.isNaN by not coercing its input. We need to polyfill it to support
+ * Internet Explorer.
```
Scratch 3.0 does not support IE, so we can take out that part of the comment.

kchadha reviewed (changes requested):
One last minor comment, and then after that LG.

kchadha reviewed (approved):
Actually, I think it's fine to leave the comment about IE as is. This PR LGTM.
